src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs

```diff
@@ -22,40 +22,105 @@ namespace Microsoft.Azure.WebJobs.Script.WebHost
+            _hostNameProvider = hostNameProvider ?? throw new ArgumentNullException(nameof(hostNameProvider));
+            _startupContextProvider = startupContextProvider ?? throw new ArgumentNullException(nameof(startupContextProvider));
+
             _repositoryClassName = _repository.GetType().Name.ToLower();
-            _hostNameProvider = hostNameProvider;
+            _repository.SecretsChanged += OnSecretsChanged;
+        }
+
+        private ConcurrentDictionary<string, IDictionary<string, string>> FunctionSecrets
+        {
+            get
+            {
+                if (_functionSecrets == null)
+                {
+                    _functionSecretsLock.Wait();
+                    try
+                    {
+                        if (_functionSecrets == null)
+                        {
+                            IDictionary<string, IDictionary<string, string>> cachedFunctionSecrets = null;
+                            if (_functionSecretsInitialLoad)
+                            {
+                                // first time secrets are accessed, check startup cache
+                                cachedFunctionSecrets = _startupContextProvider.GetFunctionSecretsOrNull();
+                                _functionSecretsInitialLoad = false;
+                            }
+                            _functionSecrets = cachedFunctionSecrets != null ?
+                                new ConcurrentDictionary<string, IDictionary<string, string>>(cachedFunctionSecrets, StringComparer.OrdinalIgnoreCase) :
+                                new ConcurrentDictionary<string, IDictionary<string, string>>(StringComparer.OrdinalIgnoreCase);
+                        }
+                    }
+                    finally
+                    {
+                        _functionSecretsLock.Release();
+                    }
+                }
+                return _functionSecrets;
+            }
 
-            if (createHostSecretsIfMissing)
+            set
             {
-                // GetHostSecrets will create host secrets if not present
-                GetHostSecretsAsync().GetAwaiter().GetResult();
+                _functionSecrets = value;
+            }
+        }
+
+        private HostSecretsInfo HostSecrets
+        {
+            get
+            {
+                if (_hostSecrets == null && _hostSecretsInitialLoad)
+                {
+                    try
+                    {
+                        // first time secrets are accessed, check startup cache
+                        _hostSecretsLock.Wait();
+                        _hostSecrets = _startupContextProvider.GetHostSecretsOrNull();
+                        _hostSecretsInitialLoad = false;
+                    }
+                    finally
+                    {
+                        _hostSecretsLock.Release();
+                    }
+                }
+                return _hostSecrets;
```

Comment:
In `FunctionSecrets` -- if it's not the initial load, you initialize it to a new Dictionary. Do you need to initialize it to a new `HostSecretsInfo`? Here it looks like if it's not the initial load, `_hostSecrets` will remain `null`.

Reply:
I had introduced all this complexity to remove the init from the ctor, due to coldstart concerns for non-auth scenarios. However, after testing, the SecretManager isn't created in those scenarios. So I've removed these properties and all this complexity, reverting back to what I had in an earlier iteration.